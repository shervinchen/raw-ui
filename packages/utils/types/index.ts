export type DeepPartial<T> = {
  [P in keyof T]?: T[P] extends Record<string, any> ? DeepPartial<T[P]> : T[P];
};

export type Merge<T, P> = P & Omit<T, keyof P>;
