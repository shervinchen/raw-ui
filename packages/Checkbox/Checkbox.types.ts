import { ChangeEvent, ComponentPropsWithoutRef } from 'react';
import { Merge } from '../utils';

export type CheckboxValue = string | number | undefined;

export interface BaseCheckboxProps {
  checked?: boolean;
  defaultChecked?: boolean;
  value?: CheckboxValue;
  disabled?: boolean;
  indeterminate?: boolean;
  className?: string;
  onChange?: (event: ChangeEvent<HTMLInputElement>) => void;
}

export type CheckboxProps = Merge<
  ComponentPropsWithoutRef<'input'>,
  BaseCheckboxProps
>;

export interface CheckboxIconProps {
  checked?: boolean;
  indeterminate?: boolean;
}
