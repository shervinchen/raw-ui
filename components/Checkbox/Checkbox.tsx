import React, {
  FC,
  ChangeEvent,
  useEffect,
  useMemo,
} from "react";
import classNames from "classnames";
import CheckboxIcon from "./CheckboxIcon";
import { CheckboxProps } from "./Checkbox.types";
import { useControlled } from "../utils/hooks";
import { useCheckboxGroupContext } from "./checkbox-group-context";

const Checkbox: FC<CheckboxProps> = ({
  defaultChecked = false,
  checked,
  value: checkboxValue,
  disabled = false,
  indeterminate = false,
  onChange,
  className = "",
  children,
  ...restProps
}) => {
  const { inGroup, groupDisabled, groupValue, onGroupChange } = useCheckboxGroupContext()
  const [internalValue, setInternalValue] = useControlled<boolean>({
    defaultValue: defaultChecked,
    value: checked,
  });
  const isDisabled = inGroup ? groupDisabled || disabled : disabled;
  const classes = classNames(
    "raw-checkbox",
    className,
  );

  const selfChecked = useMemo(() => {
    if (!inGroup) {
      return internalValue
    }
    return groupValue.includes(checkboxValue)
  }, [internalValue, inGroup, groupValue, checkboxValue])

  const changeHandler = (event: ChangeEvent<HTMLInputElement>) => {
    if (isDisabled) return
    if (!indeterminate) setInternalValue(event.target.checked)
    if (inGroup) onGroupChange?.(checkboxValue, event.target.checked)
    onChange?.(event);
  }

  return (
    <label className={classes}>
      <CheckboxIcon checked={selfChecked} indeterminate={indeterminate} />
      <input
        className="raw-checkbox-input"
        type="checkbox"
        disabled={isDisabled}
        checked={selfChecked}
        onChange={changeHandler}
        {...restProps}
      />
      {children && <span className="raw-checkbox-text">{children}</span>}
      <style jsx>{`
        .raw-checkbox {
          display: inline-flex;
          align-items: center;
          cursor: ${isDisabled ? 'not-allowed' : 'pointer'};
          opacity: ${isDisabled ? 0.5 : 1};
        }
        .raw-checkbox-input {
          position: absolute;
          opacity: 0;
          outline: none;
          width: 0;
          height: 0;
          margin: 0;
          padding: 0;
          z-index: -1;
          font-size: 0;
          background-color: transparent;
        }
        .raw-checkbox-text {
          font-size: 14px;
          line-height: normal;
          margin-left: 8px;
          user-select: none;
        }
      `}</style>
    </label>
  )
}

export default Checkbox;
