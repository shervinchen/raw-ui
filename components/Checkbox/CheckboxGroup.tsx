
import React, {
  FC,
  PropsWithChildren,
  useEffect,
  useMemo,
} from "react";
import classNames from "classnames";
import { CheckboxGroupConfig, CheckboxGroupProps, CheckboxGroupValue } from "./CheckboxGroup.types";
import { useControlled } from "../utils/hooks";
import { CheckboxGroupContext } from "./checkbox-group-context";
import { CheckboxValue } from "./Checkbox.types";

const CheckboxGroup: FC<PropsWithChildren<CheckboxGroupProps>> = ({
  value: checkboxGroupValue,
  defaultValue = [],
  disabled = false,
  className = '',
  layout = 'row',
  onChange,
  children,
  ...restProps
}) => {
  const [internalValue, setInternalValue] = useControlled<CheckboxGroupValue>({
    defaultValue,
    value: checkboxGroupValue,
  });
  const classes = classNames(
    "raw-checkbox-group",
    className,
  );

  const groupChangeHandler = (checkboxValue?: CheckboxValue, checked?: boolean) => {
    const nextGroupValue = checked ? [...internalValue, checkboxValue] : internalValue.filter(value => value !== checkboxValue)
    setInternalValue(nextGroupValue)
    onChange?.(nextGroupValue)
  }

  const checkboxGroupConfig = useMemo<CheckboxGroupConfig>(() => {
    return {
      groupValue: internalValue,
      groupDisabled: disabled,
      inGroup: true,
      onGroupChange: groupChangeHandler
    }
  }, [internalValue, disabled])

  return (
    <CheckboxGroupContext.Provider value={checkboxGroupConfig}>
      <div className={classes} {...restProps}>
        {children}
        <style jsx>{`
          .raw-checkbox-group {
            display: flex;
            flex-direction: ${layout};
            gap: 32px;
          }
        `}</style>
      </div>
    </CheckboxGroupContext.Provider>
  )
}

export default CheckboxGroup;
