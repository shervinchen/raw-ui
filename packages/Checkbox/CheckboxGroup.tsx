import React, { FC, PropsWithChildren, useCallback, useMemo } from 'react';
import classNames from 'classnames';
import {
  CheckboxGroupConfig,
  CheckboxGroupProps,
  CheckboxGroupValue,
} from './CheckboxGroup.types';
import { useControlled } from '../utils/hooks';
import { CheckboxGroupContext } from './checkbox-group-context';
import { CheckboxValue } from './Checkbox.types';

const CheckboxGroup: FC<PropsWithChildren<CheckboxGroupProps>> = ({
  value: checkboxGroupValue,
  defaultValue = [],
  disabled = false,
  className = '',
  layout = 'row',
  onChange,
  children,
  ...restProps
}) => {
  const [internalValue, setInternalValue] = useControlled<CheckboxGroupValue>({
    defaultValue,
    value: checkboxGroupValue,
  });
  const classes = classNames('raw-checkbox-group', className);

  const handleGroupChange = useCallback(
    (checkboxValue?: CheckboxValue, checked?: boolean) => {
      const nextGroupValue = checked
        ? [...internalValue, checkboxValue]
        : internalValue.filter((value) => value !== checkboxValue);
      setInternalValue(nextGroupValue);
      onChange?.(nextGroupValue);
    },
    [internalValue, onChange, setInternalValue]
  );

  const checkboxGroupConfig = useMemo<CheckboxGroupConfig>(() => {
    return {
      groupValue: internalValue,
      groupDisabled: disabled,
      inGroup: true,
      onGroupChange: handleGroupChange,
    };
  }, [internalValue, disabled, handleGroupChange]);

  return (
    <CheckboxGroupContext.Provider value={checkboxGroupConfig}>
      <div
        role="group"
        className={classes}
        {...restProps}
        data-testid="checkboxGroup"
      >
        {children}
        <style jsx>{`
          .raw-checkbox-group {
            display: inline-flex;
            flex-direction: ${layout};
            gap: 32px;
          }
        `}</style>
      </div>
    </CheckboxGroupContext.Provider>
  );
};

export default CheckboxGroup;
