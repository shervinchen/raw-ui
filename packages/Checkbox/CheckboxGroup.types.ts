import { ComponentPropsWithoutRef } from 'react';
import { CheckboxValue } from './Checkbox.types';
import { Merge } from '../utils';

export type CheckboxGroupValue = (CheckboxValue | undefined)[];

export interface BaseCheckboxGroupProps {
  defaultValue?: CheckboxGroupValue;
  value?: CheckboxGroupValue;
  disabled?: boolean;
  layout?: 'row' | 'column';
  onChange?: (value: CheckboxGroupValue) => void;
  className?: string;
}

export type CheckboxGroupProps = Merge<
  ComponentPropsWithoutRef<'div'>,
  BaseCheckboxGroupProps
>;

export interface CheckboxGroupConfig {
  groupDisabled: boolean;
  groupValue: CheckboxGroupValue;
  inGroup: boolean;
  onGroupChange?: (checkboxValue?: CheckboxValue, checked?: boolean) => void;
}
