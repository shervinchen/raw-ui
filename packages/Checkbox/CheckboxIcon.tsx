import React, { FC } from 'react';
import { Check, Minus } from 'react-feather';
import { CheckboxIconProps } from './Checkbox.types';
import { RawUITheme } from '../Theme/preset/preset.type';
import { useTheme } from '../Theme/theme-context';

const CheckboxIcon: FC<CheckboxIconProps> = ({ checked, indeterminate }) => {
  const theme: RawUITheme = useTheme();

  return (
    <span className="raw-checkbox-inner">
      <div className="raw-checkbox-icon">
        {checked && !indeterminate && (
          <Check size={12} color={theme.palette.background} strokeWidth={3} />
        )}
        {indeterminate && (
          <Minus size={12} color={theme.palette.accents7} strokeWidth={3} />
        )}
      </div>
      <style jsx>{`
        .raw-checkbox-inner {
          position: relative;
          width: 16px;
          height: 16px;
          border: 1px solid
            ${checked && !indeterminate
              ? theme.palette.foreground
              : theme.palette.accents7};
          background-color: ${checked && !indeterminate
            ? theme.palette.foreground
            : 'transparent'};
          border-radius: 3px;
        }
        .raw-checkbox-icon {
          display: inline-flex;
          justify-content: center;
          align-items: center;
          position: absolute;
          width: 16px;
          height: 16px;
          top: -1px;
          left: -1px;
        }
      `}</style>
    </span>
  );
};

export default CheckboxIcon;
