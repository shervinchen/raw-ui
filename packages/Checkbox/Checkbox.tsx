import React, {
  FC,
  ChangeEvent,
  useMemo,
  PropsWithChildren,
  useRef,
  useEffect,
  useCallback,
} from 'react';
import classNames from 'classnames';
import CheckboxIcon from './CheckboxIcon';
import { CheckboxProps } from './Checkbox.types';
import { useControlled } from '../utils/hooks';
import { useCheckboxGroupContext } from './checkbox-group-context';
import { VisuallyHiddenInput } from '../VisuallyHidden';

const Checkbox: FC<PropsWithChildren<CheckboxProps>> = ({
  defaultChecked = false,
  checked,
  value: checkboxValue,
  disabled = false,
  indeterminate = false,
  onChange,
  className = '',
  children,
  ...restProps
}) => {
  const { inGroup, groupDisabled, groupValue, onGroupChange } =
    useCheckboxGroupContext();
  const [internalValue, setInternalValue] = useControlled<boolean>({
    defaultValue: defaultChecked,
    value: checked,
  });
  const checkboxRef = useRef<HTMLInputElement>(null);
  const isDisabled = inGroup ? groupDisabled || disabled : disabled;
  const classes = classNames('raw-checkbox', className);

  const selfChecked = useMemo(() => {
    if (!inGroup) {
      return internalValue;
    }
    return groupValue.includes(checkboxValue);
  }, [internalValue, inGroup, groupValue, checkboxValue]);

  const setIndeterminate = useCallback(() => {
    const checkBox = checkboxRef.current as HTMLInputElement;
    if (indeterminate) {
      checkBox.indeterminate = true;
    } else if (checkBox.indeterminate) {
      checkBox.indeterminate = false;
    }
  }, [indeterminate]);

  const handleChange = (event: ChangeEvent<HTMLInputElement>) => {
    if (isDisabled) return;
    setIndeterminate();
    if (!indeterminate) setInternalValue(event.target.checked);
    if (inGroup) onGroupChange?.(checkboxValue, event.target.checked);
    onChange?.(event);
  };

  useEffect(() => {
    setIndeterminate();
  }, [checkboxRef, indeterminate, setIndeterminate]);

  return (
    <label className={classes} data-testid="checkboxLabel">
      <VisuallyHiddenInput
        ref={checkboxRef}
        className="raw-checkbox-input"
        type="checkbox"
        disabled={isDisabled}
        checked={selfChecked}
        onChange={handleChange}
        aria-checked={indeterminate ? 'mixed' : selfChecked ? 'true' : 'false'}
        {...restProps}
      />
      <CheckboxIcon checked={selfChecked} indeterminate={indeterminate} />
      {children && <span className="raw-checkbox-text">{children}</span>}
      <style jsx>{`
        .raw-checkbox {
          position: relative;
          display: inline-flex;
          align-items: center;
          cursor: ${isDisabled ? 'not-allowed' : 'pointer'};
          opacity: ${isDisabled ? 0.5 : 1};
          gap: 8px;
        }
        .raw-checkbox-text {
          font-size: 14px;
          line-height: normal;
          user-select: none;
        }
      `}</style>
    </label>
  );
};

Checkbox.displayName = 'RawCheckbox';

export default Checkbox;
