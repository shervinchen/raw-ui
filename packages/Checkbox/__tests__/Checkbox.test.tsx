import React, { ChangeEvent, useState } from 'react';
import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import Checkbox from '..';
import { CheckboxProps } from '../Checkbox.types';
import { CheckboxGroupProps } from '../CheckboxGroup.types';

const optionsData = [
  {
    name: 'React',
    value: 'react',
    disabled: false,
  },
  {
    name: 'Vue',
    value: 'vue',
    disabled: false,
  },
  {
    name: 'Angular',
    value: 'angular',
    disabled: false,
  },
  {
    name: 'Svelte',
    value: 'svelte',
    disabled: true,
  },
];

const user = userEvent.setup({ advanceTimers: jest.advanceTimersByTime });

describe('Checkbox', () => {
  test('should match the snapshot', () => {
    const { asFragment } = render(<Checkbox />);
    expect(asFragment()).toMatchSnapshot();
  });

  test('should support custom class name', () => {
    const { container } = render(<Checkbox className="custom-checkbox" />);
    expect(container.firstChild).toHaveClass('custom-checkbox');
  });

  test('should trigger event when clicked', async () => {
    const clickHandler = jest.fn();
    const { container } = render(<Checkbox onClick={clickHandler} />);
    const checkbox = container.firstChild;
    await user.click(checkbox as Element);
    expect(clickHandler).toHaveBeenCalledTimes(1);
  });

  test('should support uncontrolled value', () => {
    const { container } = render(<Checkbox defaultChecked />);
    const checkboxInput = container.querySelector('.raw-checkbox-input');
    expect(checkboxInput).toBeChecked();
  });

  test('should support label text', () => {
    const { queryByText } = render(<Checkbox>Label</Checkbox>);
    expect(queryByText('Label')).toBeInTheDocument();
  });

  test('should support indeterminate', () => {
    const { container } = render(
      <Checkbox indeterminate>Indeterminate</Checkbox>
    );
    const checkboxInput = container.querySelector('.raw-checkbox-input');
    expect(checkboxInput).toBePartiallyChecked();
  });

  test('should support disabled', async () => {
    const { container } = render(<Checkbox disabled />);
    const checkbox = container.firstChild;
    const checkboxInput = container.querySelector('.raw-checkbox-input');
    expect(checkboxInput).toBeDisabled();
    await user.click(checkbox as Element);
    expect(checkboxInput).not.toBeChecked();
  });

  test('should support controlled value', async () => {
    const onChange = jest.fn();

    const Component = (props: CheckboxProps) => {
      const [checked, setChecked] = useState(false);

      return (
        <Checkbox
          checked={checked}
          onChange={(event: ChangeEvent<HTMLInputElement>) => {
            setChecked(event.target.checked);
            props.onChange?.(event);
          }}
        >
          Controlled
        </Checkbox>
      );
    };
    const { container } = render(<Component onChange={onChange} />);
    const checkbox = container.firstChild;
    const checkboxInput = container.querySelector('.raw-checkbox-input');
    expect(checkboxInput).not.toBeChecked();
    await user.click(checkbox as Element);
    expect(checkboxInput).toBeChecked();
    expect(onChange).toHaveBeenCalledTimes(1);
  });

  test('should not change value when indeterminate', async () => {
    const { container } = render(<Checkbox indeterminate />);
    const checkbox = container.firstChild;
    const checkboxInput = container.querySelector('.raw-checkbox-input');
    expect(checkboxInput).toBePartiallyChecked();
    await user.click(checkbox as Element);
    expect(checkboxInput).toBePartiallyChecked();
  });

  test('should support checkbox group uncontrolled value', async () => {
    const Component = () => (
      <Checkbox.Group defaultValue={['react', 'vue']}>
        {optionsData.map((item) => (
          <Checkbox value={item.value} key={item.value}>
            {item.name}
          </Checkbox>
        ))}
      </Checkbox.Group>
    );
    const { container } = render(<Component />);
    const checkboxOne = container.querySelectorAll('input')[0];
    const checkboxTwo = container.querySelectorAll('input')[1];
    const checkboxThree = container.querySelectorAll('input')[2];
    const checkboxFour = container.querySelectorAll('input')[3];
    expect(checkboxOne).toBeChecked();
    expect(checkboxTwo).toBeChecked();
    expect(checkboxThree).not.toBeChecked();
    expect(checkboxFour).not.toBeChecked();
    await user.click(checkboxTwo);
    await user.click(checkboxThree);
    await user.click(checkboxFour);
    expect(checkboxOne).toBeChecked();
    expect(checkboxTwo).not.toBeChecked();
    expect(checkboxThree).toBeChecked();
    expect(checkboxFour).toBeChecked();
  });

  test('should support checkbox group controlled value', async () => {
    let checked = ['react', 'vue'];

    const onChange = jest.fn((value: string[]) => {
      checked = value;
    });

    const Component = (props: CheckboxGroupProps) => (
      <Checkbox.Group {...props}>
        {optionsData.map((item) => (
          <Checkbox value={item.value} key={item.value}>
            {item.name}
          </Checkbox>
        ))}
      </Checkbox.Group>
    );

    const { container, rerender } = render(
      <Component value={checked} onChange={onChange} />
    );

    const [checkboxOne, checkboxTwo, checkboxThree, checkboxFour] = Array.from(
      container.querySelectorAll('input')
    );

    expect(checkboxOne).toBeChecked();
    expect(checkboxTwo).toBeChecked();
    expect(checkboxThree).not.toBeChecked();
    expect(checkboxFour).not.toBeChecked();

    await user.click(checkboxThree);
    rerender(<Component value={checked} onChange={onChange} />);
    await user.click(checkboxFour);
    rerender(<Component value={checked} onChange={onChange} />);

    expect(onChange).toHaveBeenCalledTimes(2);
    expect(checked).toEqual(['react', 'vue', 'angular', 'svelte']);
  });

  test('should support checkbox group disabled', async () => {
    const Component = () => (
      <Checkbox.Group defaultValue={[]} disabled>
        <Checkbox value="react">React</Checkbox>
        <Checkbox value="vue">Vue</Checkbox>
        <Checkbox value="angular" disabled>
          Angular
        </Checkbox>
        <Checkbox value="svelte" disabled={false}>
          Svelte
        </Checkbox>
      </Checkbox.Group>
    );
    const { container } = render(<Component />);
    const [checkboxOne, checkboxTwo, checkboxThree, checkboxFour] = Array.from(
      container.querySelectorAll('input')
    );
    expect(checkboxOne).toBeDisabled();
    expect(checkboxTwo).toBeDisabled();
    expect(checkboxThree).toBeDisabled();
    expect(checkboxFour).toBeDisabled();

    await user.click(checkboxOne);
    await user.click(checkboxTwo);
    await user.click(checkboxThree);
    await user.click(checkboxFour);

    expect(checkboxOne).not.toBeChecked();
    expect(checkboxTwo).not.toBeChecked();
    expect(checkboxThree).not.toBeChecked();
    expect(checkboxFour).not.toBeChecked();
  });

  test('should support row and col layout in checkbox group', () => {
    const Component = ({ layout }: { layout: 'row' | 'column' }) => (
      <Checkbox.Group defaultValue={['react', 'vue']} layout={layout}>
        {optionsData.map((item) => (
          <Checkbox value={item.value} key={item.value}>
            {item.name}
          </Checkbox>
        ))}
      </Checkbox.Group>
    );
    const { container, rerender } = render(<Component layout="row" />);
    const checkboxGroup = container.firstChild as Element;
    expect(getComputedStyle(checkboxGroup).flexDirection).toBe('row');
    rerender(<Component layout="column" />);
    expect(getComputedStyle(checkboxGroup).flexDirection).toBe('column');
  });
});
