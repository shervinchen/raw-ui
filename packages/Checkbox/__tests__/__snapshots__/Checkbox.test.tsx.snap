// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox should match the snapshot 1`] = `
<DocumentFragment>
  <label
    class="jsx-4106493912 raw-checkbox"
    data-testid="checkboxLabel"
  >
    <input
      aria-checked="false"
      class="jsx-3265204931 raw-checkbox-input"
      type="checkbox"
    />
    <span
      class="jsx-3654000413 raw-checkbox-inner"
    >
      <div
        class="jsx-3654000413 raw-checkbox-icon"
      />
    </span>
  </label>
</DocumentFragment>
`;
